### ISSUE TITLE
Problem of reverse_view after filter_view

### ISSUE THREAD

**Issue by adah1972** [CaseyCarter/cmcstl2#214]
I checked with range-v3, the following code had no problem (when cmcstl2 does not conflict with range-v3):

```c++
#include <range/v3/all.hpp>
#include <experimental/ranges/algorithm>
#include <experimental/ranges/iterator>
#include <algorithm>
#include <iostream>
#include <vector>

int main()
{
    namespace std2 = std::experimental::ranges;
    using namespace ranges::v3;
    int v[] = {1, 7, 3, 6, 5, 2, 4, 8};
    auto r = v
        | view::filter([](int i) { return i % 2 == 0; })
        | view::reverse;
    copy(v, std2::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
    copy(r, std2::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
}
```

However, the equivalent std-ranges version does not work:

```c++
#include <experimental/ranges/algorithm>
#include <experimental/ranges/iterator>
#include <experimental/ranges/ranges>
#include <algorithm>
#include <iostream>
#include <vector>

int main()
{
    namespace ranges = std::experimental::ranges;
    int a[] = {1, 7, 3, 6, 5, 2, 4, 8};
    //auto r = ranges::filter_view(ranges::reverse_view(a),
    //                             [](int i) { return i % 2 == 0; });
    auto r = ranges::reverse_view(
        ranges::filter_view(a, [](int i) { return i % 2 == 0; }));
    copy(a, ranges::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
    copy(r, ranges::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
}
```

I got invalid output followed by a segment fault. The commented-out code can work without problems.

Range-v3 has no problems with either order. I am wondering whether my code is non-conformant, or there is a bug somewhere in cmcstl2?

**Comment by adah1972** [comment]
Something got messed up (the second piece of code does not compile without modifications). This is the complete code that demonstrates the problem:

```cpp
#include <experimental/ranges/algorithm>
#include <experimental/ranges/iterator>
#include <experimental/ranges/ranges>
#include <algorithm>
#include <iostream>
#include <vector>

int main()
{
    namespace ranges = std::experimental::ranges;
    int a[] = {1, 7, 3, 6, 5, 2, 4, 8};
    //auto r = ranges::filter_view(ranges::reverse_view(a),
    //                             [](int i) { return i % 2 == 0; });
    auto r = ranges::reverse_view(
        ranges::filter_view(a, [](int i) { return i % 2 == 0; }));
    ranges::copy(a, ranges::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
    ranges::copy(r, ranges::ostream_iterator<int>(std::cout, " "));
    std::cout << std::endl;
}
```

If I change the type of `a` from `int [8]` to `std::vector<int>`, I got a compile-time error instead. Interesting!

```
In file included from /usr/local/Cellar/gcc/8.2.0/include/c++/8.2.0/bits/move.h:55,
                 from /usr/local/Cellar/gcc/8.2.0/include/c++/8.2.0/bits/stl_pair.h:59,
                 from /usr/local/Cellar/gcc/8.2.0/include/c++/8.2.0/utility:70,
                 from ../cmcstl2/include/stl2/detail/algorithm/tagspec.hpp:15,
                 from ../cmcstl2/include/stl2/algorithm.hpp:16,
                 from ../cmcstl2/include/experimental/ranges/algorithm:12,
                 from test2.cpp:1:
/usr/local/Cellar/gcc/8.2.0/include/c++/8.2.0/type_traits: In substitution of 'template<class _Tp> decltype (__declval<_Tp>(0)) std::declval() [with _Tp = const std::experimental::ranges::v1::filter_view<std::experimental::ranges::v1::ref_view<std::vector<int> >, main()::<lambda(int)> >&]':
../cmcstl2/include/stl2/detail/range/access.hpp:47:9:   recursively required by substitution of 'template<class T>  requires  _NotSameAs<T, std::experimental::ranges::v1::ref_view<std::vector<int, std::allocator<int> > > > and (<requires (T& t) <valid-expression (std::experimental::ranges::v1::ref_view<std::vector<int, std::allocator<int> > >::fun()(static_cast<T&>(t)))>> constexpr std::experimental::ranges::v1::ref_view<std::vector<int> >::ref_view(T&&) [with T = std::experimental::ranges::v1::filter_view<std::experimental::ranges::v1::ref_view<std::vector<int> >, main()::<lambda(int)> >]'
../cmcstl2/include/stl2/detail/range/access.hpp:47:9:   required by substitution of 'template<class Rng> std::experimental::ranges::v1::reverse_view(Rng&&)-> std::experimental::ranges::v1::reverse_view<decltype (std::experimental::ranges::v1::view::all(declval<R>()))> [with Rng = std::experimental::ranges::v1::filter_view<std::experimental::ranges::v1::ref_view<std::vector<int> >, main()::<lambda(int)> >]'
test2.cpp:15:65:   required from here
/usr/local/Cellar/gcc/8.2.0/include/c++/8.2.0/type_traits:756:55: fatal error: template instantiation depth exceeds maximum of 900 (use -ftemplate-depth= to increase the maximum)
     auto declval() noexcept -> decltype(__declval<_Tp>(0));
                                         ~~~~~~~~~~~~~~^~~
compilation terminated.
```

**Comment by CaseyCarter** [comment]
> I am wondering whether my code is non-conformant, or there is a bug somewhere in cmcstl2?

This is almost certainly my bug and not yours.

**Comment by CaseyCarter** [comment]
> > I am wondering whether my code is non-conformant, or there is a bug somewhere in cmcstl2?
> 
> This is almost certainly my bug and not yours.

The bug is in `filter_view::__iterator::operator--`:
```diff
 constexpr __iterator& operator--() requires BidirectionalRange<V>
 {
     do
         --current_;
-    while(__stl2::invoke(parent_->pred_.get(), *current_));
+    while(!__stl2::invoke(parent_->pred_.get(), *current_));
     return *this;
 }
```
It's searching for the previous element that *doesn't* satisfy the predicate rather than the previous element that *does*. I fixed this months ago in the Ranges proposal, and assumed it was a typo that had been introduced during all the editing work. It never occurred to me that the bug might have been copied directly from the implementation.

Fixed in bf7206f7569cd7a2ca90022fb812ac79e3a79baa.

**Comment by CaseyCarter** [comment]
Thanks for the report!
